New feature add prms metadata to DataFile

pnorton-usgs opened the pull request:
This PR adds PRMS metadata (from data_file.xml) to the `DataFile` class. Some of the data file input variables specify derived units (e.g. `elev_units`, `precip_units`, `runoff_units`, and `temp_units`); a helper function `adjust_derived_units()` was added to modify the `units` attribute for those input variables to match selected units. Because the derived units parameters are defined as parameters the `Parameters` class was modified to 1) be able to adjust units for parameters with derived units (`adjust_units()` and 2) supply a dictionary of derived units parameters and their current value (`derived_units` property) that can be passed to the `DataFile.adjust_derived_units()` function.

Tests were written for this functionality.

pnorton-usgs committed:
style: renamed to data_file.xml

pnorton-usgs committed:
feat: add units values to select parameters

- added units values to elev_units, precip_units, runoff_units, and temp_units. This is to support updating units metadata for parameters that use those derived units.

pnorton-usgs committed:
doc: fix spelling

pnorton-usgs committed:
feat: add support for reading PRMS data file metadata

pnorton-usgs committed:
feat: initial support for handling derived units

- add property derived_units() which returns the unit values for elev_units, precip_units, runoff_units, and temp_units as read from a parameter file or database.
- add function adjust_units() which will adjust the units metadata value for parameters that depend on a derived units parameter value.

- bug(parameters_info): re-added missing lines from function

pnorton-usgs committed:
feat: add metadata support

- initial support for using official data file metadata
- added __str__ for pretty-print

pnorton-usgs committed:
feat: add metadata support

- initial support for using official data file metadata
- added adjust_derived_units() to modify the metadata units value for input variables that rely on the derived units parameter elev_units, precip_units, runoff_units, and temp_units for their units value.
- renamed _add_metadata() to _add_file_metadata() to reduce confusion between the file-based metadata and the official metadata.

pnorton-usgs committed:
fix: rename units to file_units

pnorton-usgs committed:
feat: add additional tests

- add test for __str__ and derived_units

pnorton-usgs committed:
style: format and typing updates

pnorton-usgs committed:
style: updates to verbose output

pnorton-usgs committed:
style: updates to verbose output

pnorton-usgs committed:
feat: add test for diff() function

jmccreight commented:
The overall idea and need for treating (potentially) user-supplied units in both `DataFile` and `Parameters` makes sense to me. However, I do not have much comment on what's going on in this PR mechanically, because I'm not very familiar how things works.

One thing I dont understand: How does `DataFile` get the units from the parameter file? It would seem that a parameter object or parameter file would be a required input for `DataFile`. The design seems to be that the user loads as parameter file to a `Parameters` object and passes `parameters.metadata` to `DataFile`. Is that right? Why not just pass the parameters to DataFile (either as a file name or a `Parameters` instance)?

My only other comment is that I do not like the terminology adopted in the PR and I dont see it in the PRMS5.2.1 tables, so it seems like it could be changed. The terms "derived units" and "adjust_units" are both confusing/vague to me. The `Parameters.adjust_units()` docstring made me feel very sad that I had no idea what it was doing. LOL. The terms could be more specific like "input units" and "resolve_input_units"/"apply_input_units", this would really help the user perceive what is going on. Names are about the biggest deal in programming.

pnorton-usgs committed:
refactor: change adjust_derived_units to resolve_defined_units

pnorton-usgs committed:
refactor: change naming for defined units

- change derived_units() to defined_units()
- change adjust_units() to resolve_defined_units()

pnorton-usgs committed:
refactor: change function call for defined units

pnorton-usgs committed:
refactor: update naming and handling of user defined units

- add Parameters object argument to __init__()
- automatically apply user defined units on file read or output warning if no parameter file was provided
- rename resolve_defined_units() to resolve_units()

pnorton-usgs committed:
test: update tests for DataFile

- comment out test_datafile_parameter_units()
- add parameter file to test_read_datafile_sagehen()

pnorton-usgs committed:
fix: add missing handle for expanding nhru parameter to nhru,nmonth

- data setter was missing handling for expanding a parameter with nhru dimension to nhru x nmonth
- add verbose argument to class
- add rich library

pnorton-usgs committed:
refactor: updates for handling user defined units parameters

- rename defined_units() property to user_defined_units()
- add warning output to user_defined_units() when an expected unit parameter is missing
- rename resolve_defined_units() to resolve_units()
- the check() routine was not correctly reporting all cases where values for a parameter were all equal. Now it reports if 1) all values are the same, 2) all values by row are the same, and 3) all values by column are the same.

pnorton-usgs committed:
style: misc formatting

pnorton-usgs committed:
test: add parameter file for DataFile tests
